File: elixir/nucleotide-count/dna.exs
defmodule DNA do
  @nucleotides [?A, ?T, ?C, ?G]

  def count(chain, nucleotide) do
    Enum.count(chain, &1 == nucleotide)
  end

  def nucleotide_counts(chain) do
    HashDict.new @nucleotides, fn(nucleotide) ->
      {nucleotide, count(chain, nucleotide)}
    end
  end
end

File: scala/nucleotide-count/src/main/scala/dna.scala
class DNA(strand: String) {
  require(DNA.validNucleotideStrand(strand))

  def count(nucleotide: Char) = {
    require(DNA.validNucleotide(nucleotide) || nucleotide == 'U')

    nucleotideCounts.getOrElse(nucleotide, 0)
  }

  lazy val nucleotideCounts =
    DNA.nucleotides.map { nucleotide =>
      (nucleotide -> strand.count(_ == nucleotide))
    }.toMap
}

object DNA {
  private val nucleotides = Set('A', 'T', 'G', 'C')

  private def validNucleotide(nucleotide: Char) =
    nucleotides.contains(nucleotide)

  private def validNucleotideStrand(strand: String) =
    strand.forall { validNucleotide(_) }
}
